src/toil/cwl/cwltoil.py

```diff
@@ -792,25 +1193,217 @@ def write_file(writeFunc: Any, index: dict, existing: dict, file_uri: str) -> st
                 raise
         return index[file_uri]
 
+def path_to_loc(obj: Dict) -> None:
+    """
+    If a CWL object has a "path" and not a "location", make the
+    path into a location instead.
+    """
+    if "location" not in obj and "path" in obj:
+        obj["location"] = obj["path"]
+        del obj["path"]
+
+def import_files(
+    import_function: Callable[[str], FileID],
+    fs_access: cwltool.stdfsaccess.StdFsAccess,
+    fileindex: Dict,
+    existing: Dict,
+    cwl_object: Dict,
+    skip_broken: bool = False
+) -> None:
+    """
+    From the leader or worker, prepare all files and directories inside the
+    given CWL tool, order, or output object to be used on the workers. Make
+    sure their sizes are set and import all the files.
+
+    Recurses inside directories using the fs_access to find files to upload and
+    subdirectory structure to encode, even if their listings are not set or not
+    recursive.
+
+    Preserves any listing fields.
+
+    If a file cannot be found (like if it is an optional secondary file that doesn't exist),
+
+    Also does some miscelaneous normalization.
+
+    :param import_function: The function used to upload a file:// URI and get a
+    Toil FileID for it.
+
+    :param fs_access: the CWL FS access object we use to access the filesystem
+    to find files to import.
+
+    :param fileindex: Forward map to fill in from file URI to Toil storage
+    location, used by write_file to deduplicate writes.
+
+    :param existing: Reverse map to fill in from Toil storage location to file
+    URI. Not read from.
+
+    :param cwl_object: CWL tool (or workflow order) we are importing files for
+
+    :param skip_broken: If True, when files can't be imported because they e.g.
+    don't exist, leave their locations alone rather than failing with an error.
+    """
+
+    try:
+        tool_id = cwl_object['id']
+    except:
+        tool_id = str(cwl_object)
```

Comment:
Suggestion:

```
    try:
        tool_id = cwl_object['id']
    except:
        tool_id = str(cwl_object)
```

->

```
tool_id = cwl_object.get('id', str(cwl_object))
```

Reply:
Yeah, this will work as long as the type constraints are followed. I think I meant to also maybe support lists?

Reply:
Although it will always evaluate `str()`